QuestionAnswering/QA_Agent/deployment/fastapi_app/services/nlq2action/execute_action/func/sg_carpark.py

```diff
@@ -0,0 +1,103 @@
+import logging
+import time
+from typing import Annotated, List
+
+from fastapi import Depends
+
+from services.support_data import QAStep
+from core.kg import KgClient
+from utils.rdf import flatten_sparql_response
+from services.kg import get_sgCarpark_bgClient
+from services.feature_info_client import FeatureInfoClient, get_featureInfoClient
+from services.geocoding import IGeocoder, get_geocoder
+from services.support_data import TableDataItem
+from services.nlq2action.execute_action.func.base import Name2Func
+
+logger = logging.getLogger(__name__)
+
+
+class SGCarParkFuncExecutor(Name2Func):
+    def __init__(
+        self,
+        bg_client: KgClient,
+        feature_info_client: FeatureInfoClient,
+        geocoder: IGeocoder,
+    ):
+        self.bg_client = bg_client
+        self.feature_info_client = feature_info_client
+        self.geocoder = geocoder
+
+    def get_name2func(self):
+        return {"find_nearest_carpark": self.find_nearest_carpark}
+
+    def find_nearest_carpark(self, location: str):
+        steps: List[QAStep] = []
+
+        logger.info("Get coordinates for the location: " + location)
+        timestamp = time.time()
+        place = self.geocoder.search(location)
+        latency = time.time() - timestamp
+        logger.info("Geo-decoded data: " + str(place))
+        steps.append(
+            QAStep(
+                action="geodecode",
+                arguments=location,
+                results=str(place),
+                latency=latency,
+            )
+        )
+
+        timestamp = time.time()
+        query = """PREFIX carpark:	<https://www.theworldavatar.com/kg/ontocarpark/>
+    PREFIX geo: <http://www.opengis.net/ont/geosparql#>
+    PREFIX geof: <http://www.opengis.net/def/function/geosparql/>
+    PREFIX unit: <http://qudt.org/vocab/unit/>
+            
+    SELECT * WHERE {{
+    ?Carpark a carpark:Carpark; rdfs:label ?Label .
+    SERVICE <http://sg-ontop:8080/sparql> {{
```

Comment:
just a comment, don't think it's a good idea to hardcode this, if we spin up the stack with a different stack name this breaks

Reply:
Is the endpoint http://sg-ontop:8080/sparql (which I suppose is referencing the local network of the SG stack) the same as http://174.138.23.221:3838/ontop/ui/sparql? I already configure the latter as an environment variable, so if the 2 endpoints are the same, then I'll interpolate the query string with the latter. Otherwise, I'll define a new environment variable for this.

Reply:
No, they are not the same, one is for accessing within the stack and one is from outside

Reply:
A possible solution is to have an access agent within the stack which you send queries to, instead of sending it directly to blazegraph which will figure out the correct ontop endpoint

Reply:
I tested the query with http://sg-ontop:8080/sparql and http://174.138.23.221:3838/ontop/ui/sparql and it seems that they give the same results/behaviours? Does the difference lie only in speed? In other words, is the use of either endpoint correct?

> A possible solution is to have an access agent within the stack which you send queries to, instead of sending it directly to blazegraph which will figure out the correct ontop endpoint

Yes, this will ensure that Zaha doesn't need to know the internal endpoints of the SG stack.

Reply:
> A possible solution is to have an access agent within the stack which you send queries to, instead of sending it directly to blazegraph which will figure out the correct ontop endpoint

Will this be something that you plan to implement within next week?

Reply:
> > A possible solution is to have an access agent within the stack which you send queries to, instead of sending it directly to blazegraph which will figure out the correct ontop endpoint
> 
> Will this be something that you plan to implement within next week or so?

It's not straightforward to implement this for the Sg stack/dataset, can explain more to you in person

Reply:
if this is not in your work plan in the next 1-2 weeks then for immediate action I'll use a environment variable for `http://sg-ontop:8080/sparql`.